core/contractsdk/java/example/erc20/src/main/java/com/baidu/xuper/Erc20.java

```diff
@@ -0,0 +1,323 @@
+        if (fromBalanceByte == null) {
+            return Response.error("no from found");
+        }
+        BigInteger fromBalance = new BigInteger(fromBalanceByte);
+        if (fromBalance.compareTo(tokenAmount) == -1){
+            return Response.error("no enough balance");
+        }
+
+        String toKey = BALANCEPRE + to;
+        byte[] toBalanceByte = ctx.getObject(toKey.getBytes());
+        BigInteger toBalance;
+        if (toBalanceByte == null) {
+            toBalance = BigInteger.valueOf(0);
+        } else {
+            toBalance = new BigInteger(toBalanceByte);
+        }
+
+        BigInteger fromBalanceNow = fromBalance.subtract(tokenAmount);
+        BigInteger toBalanceNow = toBalance.add(tokenAmount);
+
+        ctx.putObject(fromKey.getBytes(), fromBalanceNow.toByteArray());
+        ctx.putObject(toKey.getBytes(), toBalanceNow.toByteArray());
+
+        return Response.ok("transfer success".getBytes());
+    }
+
+    @ContractMethod
+    public Response transferFrom(Context ctx) {
+        String caller = ctx.caller();
+        if (caller.isEmpty()){
+            return Response.error("missing caller");
+        }
+
+        byte[] fromByte = ctx.args().get("from");
+        if (fromByte == null) {
+            return Response.error("missing from");
+        }
+        String from = new String(fromByte);
+
+        byte[] toByte = ctx.args().get("to");
+        if (toByte == null) {
+            return Response.error("missing to");
+        }
+        String to = new String(toByte);
+
+        byte[] tokenByte = ctx.args().get("token");
+        if (tokenByte == null) {
+            return Response.error("missing token");
+        }
+        String tokenAmountStr = new String(tokenByte);
+        BigInteger tokenAmount = new BigInteger(tokenAmountStr);
+
+        String allowanceKey = ALLOWANCEPRE + from + "_" + caller;
+        BigInteger allowance;
+        byte[] value = ctx.getObject(allowanceKey.getBytes());
+        if (value != null) {
+            allowance = new BigInteger(value);
+            if (allowance.compareTo(tokenAmount) == -1){
+                return Response.error("The allowance is not enough");
+            }
+        } else {
+            return Response.error("key " + allowanceKey + " not found");
+        }
+
+        String fromKey = BALANCEPRE + from;
+        byte[] fromBalanceByte = ctx.getObject(fromKey.getBytes());
+        if (fromBalanceByte == null) {
+            return Response.error("no from found");
+        }
+        BigInteger fromBalance = new BigInteger(fromBalanceByte);
+
+        String toKey = BALANCEPRE + to;
+        byte[] toBalanceByte = ctx.getObject(toKey.getBytes());
+        BigInteger toBalance;
+        if (toBalanceByte == null) {
+            toBalance = BigInteger.valueOf(0);
+        } else {
+            toBalance = new BigInteger(toBalanceByte);
+        }
+
+        BigInteger fromBalanceNow = fromBalance.subtract(tokenAmount);
```

Comment:
The from party also needs to check whether the balance is sufficient

Reply:
fixed